dwallet/src/eth_state.rs

```diff
@@ -0,0 +1,641 @@
+            rpc: String::default(),
+            previous_max_active_participants: u64::default(),
+            current_max_active_participants: u64::default(),
+            network: Network::default(),
+            latest_header: Header::default(),
+            last_update_execution_block_number: u64::default(),
+            last_update_execution_state_root: Bytes32::default(),
+        }
+    }
+
+    /// Sets the checkpoint for the Ethereum state.
+    ///
+    /// Set the last known checkpoint for the Ethereum state. The checkpoint
+    /// is a string that represents a specific point in the blockchain history, typically a
+    /// beacon block hash.
+    pub fn set_checkpoint(&mut self, checkpoint: String) -> Self {
+        self.last_checkpoint = checkpoint;
+        self.clone()
+    }
+
+    /// Sets the network for the Ethereum state.
+    ///
+    /// Set the network for the Ethereum state. The network is an enum that
+    /// represents the specific Ethereum network (e.g., Mainnet, Holesky, devnet, etc.).
+    pub fn set_network(&mut self, network: Network) -> Self {
+        self.network = network;
+        self.clone()
+    }
+
+    /// Sets the RPC endpoint for the Ethereum state.
+    ///
+    /// Set the RPC endpoint for the Ethereum state.
+    /// The RPC endpoint is a string that represents the URL of the Ethereum node that the client
+    /// will connect to.
+    pub fn set_rpc(&mut self, rpc: String) -> Self {
+        self.rpc = rpc;
+        self.clone()
+    }
+
+    /// Synchronizes the local state with the blockchain state based on a given checkpoint.
+    /// Performs a multi-step process to ensure the local state is up-to-date with
+    /// the blockchain's state.
+    ///
+    /// # Arguments
+    /// * `checkpoint`: A `&str` slice that represents the checkpoint from which to start the
+    ///   synchronization process. Typically, this would be a block hash or a similar identifier
+    ///   that marks a specific point in the blockchain history.
+    ///
+    /// # Process
+    ///
+    /// 1. **Bootstrap:** Initializes the synchronization process using the provided checkpoint.
+    ///    This step involves setting up the local state to match the state at the checkpoint.
+    ///
+    /// 2. **Fetch Updates:** Retrieves updates from the blockchain for the current period. The
+    ///    current period is calculated based on the slot of the last finalized header.
+    ///
+    /// 3. **Verify and Apply Updates:**
+    ///    - For each update fetched, it first verifies the update for correctness and then applies
+    ///      the update to the local state.
+    ///    - Verifies and applies a finality update, which includes updates that have been finalized
+    ///      and are irreversible.
+    ///    - Verifies and applies an optimistic update, which might still be subject to change but
+    ///      is accepted optimistically to keep the state as current as possible.
```

Comment:
```suggestion
    /// Synchronizes the local state with the blockchain state based on a given checkpoint.
    /// Performs a multistep process to ensure the local state is up-to-date with
    /// the blockchain's state.
    ///
    /// # Arguments
    /// * `checkpoint`: A `&str` slice that represents the checkpoint from which to start the
    ///   synchronization process. Typically, this would be a block hash, or a similar identifier
    ///   that marks a specific point in the blockchain history.
    ///
    /// # Process
    ///
    /// 1. **Bootstrap:** Initializes the synchronization process using the provided checkpoint.
    ///    This step involves setting up the local state to match the state at the checkpoint.
    ///
    /// 2. **Fetch Updates:** Retrieves updates from the blockchain for the current period. The
    ///    current period is calculated based on the slot of the last finalized header.
    ///
    /// 3. **Verify and Apply Updates:** – For each update fetched, it first verifies the update for
    ///    correctness and then applies the update to the local state. — Verifies and applies a
    ///    finality update, which includes updates that have been finalized and are irreversible. —
    ///    Verifies and applies an optimistic update, which might still be subject to change but is
    ///    accepted optimistically to keep the state as current as possible.
```